### ISSUE TITLE
 Functional components should support the "components" option

### ISSUE THREAD

**Issue by tatsuhirosatou** [vuejs/vue#6872]
### What problem does this feature solve?
Functional components with templates allow users to improve performance without sacrificing readability or ease of use. However, currently these do not support the "components" option. This limits this feature a lot, to a very limited set of use cases. 

### What does the proposed API look like?
Simply support the components option just like non-functional components

**Comment by c01nd01r** [comment]
Also, how to call methods and emit events in functional components with templates?

**Comment by posva** [comment]
You cannot, you use a regular component if you need that

**Comment by scottbedard** [comment]
Support for local directives would be awesome!

And @c01nd01r if you need to respond to an action, you can use plain old functions

```js
// event handler
function onClick(context, e) {
    console.log('you clicked the div');

    // to emit an event, look at the listeners in your context
    if (context.listeners.click) {
        context.listeners.click(e);
    }
}

// component
export default {
    functional: true,
    render(h, context) {
        return <div onClick={e => onClick(context, e)}>click me</div>;
    },
};
```

**Comment by c01nd01r** [comment]
Yes, @scottbedard, thank you. I know about this, but I can't do it in a "functional" html template.
```vue
<template functional>
  <div>
    <button @click="$emit('click')">Emit click</button> <!-- error -->
    <button @click="logClick">Log click</button> <!-- ??? -->
  </div>
</template>

<script>
  import { logClick } from 'logg';
</script>
```

**Comment by scottbedard** [comment]
Yea, that would be great. Right now we can't really expose things to our template in a clean way. The only technique I thought of, was a prop with a default value. This works, but it looks terrible in my opinion, and unnecessary bloats the component's API.

**Comment by miljan-aleksic** [comment]
@yyx990803, are there any plans ahead about supporting local directives with functional components? It would allow, among others, to convert some more advanced components as functional where those features could be replaced by directives.

I assume there are some technical limitations about it but would like to know if it's planned, what it requires, etc. So I can at least know why I must keep using normal components, for now.

**Comment by hirokiosame** [comment]
I opened a PR for supporting `components` in functional-components (#8143). I opened a separate issue regarding making methods accessible in functional-components (#8142); can open a PR if that gets approved.

**Comment by dougalg** [comment]
What is holding up the PR? It's been almost 4 months since it was opened.

**Comment by miljan-aleksic** [comment]
I would love to see this merged.

**Comment by cdaly7** [comment]
Any further update on this improvement?

**Comment by Fnxxxxo** [comment]
It seems the PR is blocked since of conflicts in options interface.

**Comment by kahirokunn** [comment]
Do you have new news?
I really want to use components option with functional component.

**Comment by privatenumber** [comment]
@kahirokunn 

This will likely never get approved. In v3, functional components are getting stripped of features even more in favor of stateful components supporting everything functional components can do; basically, you'll be able to update your functional component to be stateful and you'll rarely have a need for functional components in the future.

Read the RFC [here](https://github.com/vuejs/rfcs/pull/27)
